### PR TITLE
[Issue-3414] Improve performance upon showing Marketing Campaign

### PR DESCRIPTION
**Pull request by Quangdm-cdm**
Pr for issue #3414

### CODE REVIEW

**Comment by saltict** [comment]
🚀 Deployed on https://pr-3421--sw-web-runner.netlify.app

File: packages/extension-koni-ui/src/hooks/static-content/useHandleAppBannerMap.ts [hunk 1]
```diff
@@ -53,19 +30,24 @@ export const useHandleAppBannerMap = (
...
+    });
 
     dispatch(updateBannerHistoryData(result));
+  }, [bannerHistoryMap, dispatch]);
+
+  useEffect(() => {
+    initBannerHistoryMap(appBannerData);
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
+  }, [appBannerData]);
```
Comment (lw-cdm) [inline packages/extension-koni-ui/src/hooks/static-content/useHandleAppBannerMap.ts:49]: Please recheck this dependency

File: packages/extension-koni-ui/src/hooks/static-content/useHandleAppConfirmationMap.ts [hunk 2]
```diff
@@ -45,19 +30,24 @@ export const useHandleAppConfirmationMap = (
...
-    (data: AppConfirmationData[]) => {
-      getFilteredAppConfirmationByTimeAndPlatform(data);
-      initConfirmationHistoryMap(data);
-    },
-    [getFilteredAppConfirmationByTimeAndPlatform, initConfirmationHistoryMap]
-  );
+  useEffect(() => {
+    initConfirmationHistoryMap(appConfirmationData);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [appConfirmationData]);
```
Comment (lw-cdm) [inline packages/extension-koni-ui/src/hooks/static-content/useHandleAppConfirmationMap.ts:50]: Please recheck this dependency

File: packages/extension-koni-ui/src/hooks/static-content/useHandleAppPopupMap.ts [hunk 3]
```diff
@@ -53,19 +30,23 @@ export const useHandleAppPopupMap = (
...
-    (data: AppPopupData[]) => {
-      getFilteredAppPopupByTimeAndPlatform(data);
-      initPopupHistoryMap(data);
-    },
-    [getFilteredAppPopupByTimeAndPlatform, initPopupHistoryMap]
-  );
+  useEffect(() => {
+    initPopupHistoryMap(appPopupData);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [appPopupData]);
```
Comment (lw-cdm) [inline packages/extension-koni-ui/src/hooks/static-content/useHandleAppPopupMap.ts:49]: Please recheck this dependency

File: packages/extension-koni-ui/src/messaging/campaigns/index.ts [hunk 4]
```diff
@@ -17,3 +18,15 @@ export async function unlockDotCheckCanMint (request: RequestUnlockDotCheckCanMi
 export async function toggleCampaignPopup (request: ShowCampaignPopupRequest): Promise<null> {
   return sendMessage('pri(campaign.popup.toggle)', request);
 }
+
```
Comment (lw-cdm) [inline packages/extension-koni-ui/src/messaging/campaigns/index.ts:21]: If there is no usage for these functions, please remove them

**Review by nampc1 — CHANGES_REQUESTED** [review]

File: packages/extension-base/src/services/mkt-campaign-service/index.ts [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+    });
+  }
+
+  private checkCampaignExistTime (info?: AppBasicInfoData) {
+    if (info?.start_time && info?.stop_time) {
+      const now = new Date();
+      const startTime = new Date(info.start_time);
+      const endTime = new Date(info.stop_time);
+
+      if (now >= startTime && now <= endTime) {
```
[the remark this answers is not in the document]
Comment (Quangdm-cdm) [reply packages/extension-base/src/services/mkt-campaign-service/index.ts:143]: Resolved

File: packages/extension-base/src/services/mkt-campaign-service/index.ts [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+        return true;
+      } else {
+        return false;
```
[the remark this answers is not in the document]
Comment (Quangdm-cdm) [reply packages/extension-base/src/services/mkt-campaign-service/index.ts:146]: I make some change in checkCampaignExistTime function o resolve this.

File: packages/extension-base/src/services/storage-service/db-stores/YieldPositionStore.ts [hunk 7]
```diff
@@ -68,4 +68,12 @@ export default class YieldPositionStore extends BaseStore<YieldPositionInfo> {
       () => this.getByAddress(addresses)
     );
   }
+
+  async checkPositionByPoolSlug (slug: string, filterFunc: (i: YieldPositionInfo) => boolean) {
+    return this.table.where('slug').anyOfIgnoreCase(slug).filter(filterFunc).count();
```
[the remark this answers is not in the document]
Comment (Quangdm-cdm) [reply packages/extension-base/src/services/storage-service/db-stores/YieldPositionStore.ts:73]: use limit beforce count

File: packages/extension-base/src/services/storage-service/DatabaseService.ts [hunk 8]
```diff
@@ -90,6 +91,150 @@ export default class DatabaseService {
...
+    const filterFunc = (item: IBalance): boolean => {
+      const freeBalance = item?.free;
+      const lockedBalance = item?.locked;
+      const value = new BigN(freeBalance).plus(lockedBalance);
+
+      if (!addresses.includes(item.address)) {
+        return false;
+      }
+
+      switch (comparison) {
```
[the remark this answers is not in the document]
Comment (Quangdm-cdm) [reply packages/extension-base/src/services/storage-service/DatabaseService.ts:104]: resolved

File: packages/extension-base/src/services/mkt-campaign-service/index.ts [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+      data.conditions?.[type]?.forEach((condition) => {
+        if (!result[type].includes(JSON.stringify(condition))) {
+          result[type].push(JSON.stringify(condition));
+        }
+      });
+    };
+
+    respAppPopupData.forEach((popup) => {
+      if (Object.keys(popup.conditions) && Object.keys(popup.conditions).length) {
+        Object.keys(popup.conditions).forEach((con) => getConditionByType(popup, con as MktCampaignConditionTypeValue));
```
[the remark this answers is not in the document]
Comment (Quangdm-cdm) [reply packages/extension-base/src/services/mkt-campaign-service/index.ts]: I have refactored this code to make it easier to understand

**Review by nampc1 — APPROVED** [review]

**Pull request merged**